### PR TITLE
feat: Download CSV for Camp Session

### PR DESCRIPTION
**Pull request by Jasraj-Dhanoa**
## Notion ticket link

[Camp Overview: Trigger CSV functionality](https://www.notion.so/uwblueprintexecs/Camp-Overview-Trigger-CSV-functionality-86ffe05420a84280a9c8bcec77606dd7)

## Implementation description
* Added new function to CampersAPIClients to call backend API for getting CSV string corresponding to camp session ID
* Passed down camp session ID from CampersTable to CamperTable, which is then passed to export CSV button 
* When the export CSV button is clicked, the function in CampersAPIClients is called and the returned CSV string is converted and then downloaded

## Steps to test
1. Go to http://localhost:3000/admin/camp/63139c7bc3d7b55b44a01531
2. Click the "export as .csv" button
3. CSV should download and contain content corresponding to the current camp session

## What should reviewers focus on?
* Want to ensure workflow to fetch, convert and download CSV is as efficient as possible

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/services/interfaces/campService.ts [hunk 1]
```diff
@@ -54,11 +54,11 @@ interface ICampService {
 
   /**
    * Get all campers associated with camps of id campId
```
Comment (e-wai) [inline backend/typescript/services/interfaces/campService.ts:56]: Can we update this comment too please?

File: frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx [hunk 2]
```diff
@@ -30,7 +30,16 @@ import textStyles from "../../../../theme/textStyles";
...
+import CampsAPIClient from "../../../../APIClients/CampsAPIClient";
+
+import { downloadCSV } from "../../../../utils/CSVUtils";
+
+const generateCsv = async (campSessionId: string) => {
+  const csvResponse = await CampsAPIClient.getCampSessionCsv(campSessionId);
+  downloadCSV(csvResponse, "Campers_Information");
+}
+
+const ExportButton = ({campSessionId,}: {campSessionId: string}): JSX.Element => {
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx:42]: nit: Unnecessary comma after `campSessionId`

**Review by e-wai — CHANGES_REQUESTED** [review]
Can we check linter once more? Looks good to me with some small nits, will take a sec to test the endpoint and double-check the format of CSV

**Comment by e-wai** [comment]
Getting some errors testing this locally:
- The camp you mentioned is returning an error, 
```
Failed to get campers. Reason = Cast to ObjectId failed for value \"question test\" (type string) at path \"_id\" for model \"FormQuestion\""
```
- Can you try out http://localhost:3000/admin/camp/630a63450eb59852f9d01b1e for me? It's failing with a 500 server error, camp photo can't be found
```
{"message":"[/app/services/implementations/fileStorageService.ts] Failed to retrieve file. Reason = File name 1a4bf669-9b80-444c-9be3-b0ee0b9f92ec does not exist","level":"error","timestamp":"2022-10-08T08:39:56.194Z"}
scv2_ts_backend | {"message":"[/app/services/implementations/campService.ts] Failed to get camp photo for camp with id 630a63450eb59852f9d01b1e. Reason = File name 1a4bf669-9b80-444c-9be3-b0ee0b9f92ec does not exist","level":"error","timestamp":"2022-10-08T08:39:56.199Z"}
scv2_ts_backend | {"message":"[/app/services/implementations/campService.ts] Failed to get campers. Reason = Cannot read property 'keys' of undefined","level":"error","timestamp":"2022-10-08T08:39:58.658Z"}
scv2_ts_backend | {"message":"[/app/services/implementations/campService.ts] Failed to generate CSV. Reason = Cannot read property 'keys' of undefined","level":"error","timestamp":"2022-10-08T08:39:58.659Z"}
```

I suspect this is because of our inconsistent camp dev db data, but it'd be good to handle the error earlier rather than later. I'm thinking that it should not download a CSV if the GET request returns an error, in which case we can have some toast pop up with an error message. (But would be good to check all of this with product)

The happy path seems OK, but I know there's some ongoing discussion about the format of the CSV - once we sort out the errors, I'd be happy to merge this and fix it up later once we clear that up.

File: backend/typescript/services/interfaces/campService.ts (hunk 1 above)
Comment (Jasraj-Dhanoa) [reply backend/typescript/services/interfaces/campService.ts]: Done!

**Comment by Jasraj-Dhanoa** [comment]
The toast error message is now being shown when API call for CSV is unsuccessful. The 500 error occurred before due to inconsistencies in the database.

The errors "failed to retrieve file" and "failed to get camp photo" are related to the camp and according to my investigation, occur as a photo file corresponding to a camp does not exist.

File: frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx [hunk 3]
```diff
@@ -178,7 +204,7 @@ const CampersTable = ({
                 </Tag>
               );
             })}
-            <ExportButton />
+            <ExportButton campSessionId={campSessionId} campSessionCSVFilename={campSessionCSVFilename}/>
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx:207]: What are your thoughts on moving the `generateCSV()` function up to the `index.tsx` file? that way we don't have to pass information through, it is the responsibility of the parent function

**Review by e-wai — CHANGES_REQUESTED** [review]
visual nit: can we add a 40px (ish?) left margin?

Screen Shot 2022-11-11 at 5 28 07 AM

Screen Shot 2022-11-11 at 5 32 07 AM

File: frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx (hunk 3 above)
Comment (Jasraj-Dhanoa) [reply frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx]: Seems like a better option - I've made the according changes.

File: frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx [hunk 4]
```diff
@@ -33,22 +34,26 @@ import ViewCamperModal from "../ViewCamperModal/index";
...
+        aria-label="Export SVG"
+        type="submit"
+        background="background.grey.200"
+        color="primary.green.100"
+        border="2px"
+        borderColor="primary.green.100"
+        borderRadius="5px"
+        minWidth="-webkit-fit-content"
+        fontSize={textStyles.bodyRegular.fontSize}
+        onClick={() => generateCsv()}
```
Comment (e-wai) [inline frontend/src/components/pages/CampOverview/CampersTable/CampersTable.tsx:52]: nit: does `onClick={generateCSV}` work?

**Review by e-wai — APPROVED** [review]
Looks fantastic, thanks for making the changes 🎉  (left one small nit, not blocking)

**Pull request merged**